lib/ai/model/agent.go

```diff
@@ -327,7 +322,7 @@ func (a *Agent) constructScratchpad(intermediateSteps []AgentAction, observation
 // parseJSONFromModel parses a JSON object from the model output and attempts to sanitize contaminant text
 // to avoid triggering self-correction due to some natural language being bundled with the JSON.
 // The output type is generic, and thus the structure of the expected JSON varies depending on T.
-func parseJSONFromModel[T any](text string) (T, *invalidOutputError) {
```

Comment:
Is this intentional? I think the best practice is to always return `error` and let the caller cast if necessary.

Reply:
I added this some time ago, this is intentional. Maybe that's best practice but if we have one error type and it's called from 1-2 places privately, I really think casting to error is detrimental to working effectively with the code since we just throw away information and forces ourselves to write boilerplate for handling the other case and remember to update without compiler errors when we have new special types etc.

Reply:
I changed the function signature because this messes with the error handling later

<img width="1152" alt="image" src="https://github.com/gravitational/teleport/assets/16366487/0d062f3e-a002-474f-a975-f47fbfb2f2f8">

Without the signature change, it considers that `err != nil` even if the error is `nil` with the type `*model.invalidOutputError`.

Reply:
Please find a snippet demonstrating what happened: https://go.dev/play/p/zAWt-fVSmqt

Reply:
Here's a playground example demonstrating why this is a footgun: https://go.dev/play/p/VFrmccsEA8c

By implicitly setting the `err` type to `error` a few lines before, I changed the behaviour of the error check and either had to change the function signature, or store the returned error in a new variable whose type is not `error`. I preferred the first solution as this is a footgun and it took me 20 minutes to understand why this was happening.